### PR TITLE
Add list of forbidden syscalls to WASM executor

### LINKED ISSUE #587: Improve the gas spent calculation routines
### File Location(s)

https://github.com/gear-tech/gear/blob/master/pallets/gear/src/lib.rs#L305

### Proposal

- [x] Make `get_gas_spent` calculate the gas spent without affecting the message queue, it should return the gas value burned by the current message processing plus the gas value to be sent to another actor (if any).
- [x] Add the internal (mock) function to calculate the gas amount burned during message processing for using it in tests.
- [x] Calculate gas spent for init messages.
- [x] Calculate gas spent for reply messages.
- [x] Calculate gas spent for async programs.
- [x] Find the solution against the node abusing while sending an infinite program to `get_init_gas_spent()`.
- [ ] Fix the problem when a program uses `gas_available()` while sending messages.

### PR DESCRIPTION
**Pull request by shamilsan**
Finally resolves #587.

- Add functions (aka syscalls) blacklist.
- Generate a trap when trying to call a forbidden function in a program.
- Forbid to call `gr_gas_available` when executing WASM from `get_gas_spent` function.

TODO:

- [x] Fix tests 

### Use case

Now it is possible to pass a list of forbidden syscalls when calling the `gear_core_processor::process()` function. If the WASM program contains a call to that forbidden function, the execution will return a correspondent error.

For example when we calculate the amount of gas needed for processing messages in the `get_gas_spent()` function, we forbid calling the `gr_gas_available` syscall in the program. `get_gas_spent()` will return `Program terminated with a trap: Unable to call a forbidden function` error if the program calls `exec:gas_available()` function during message processing.

Up to now, the main end-user application is that he will receive an error when trying to calculate the gas amount to be spent for handling a message by the program that uses the `exec::gas_available()` function inside.

Also, there are potential use cases in the future.

@gear-tech/dev

### CODE REVIEW

**Comment by shamilsan** [comment]
> Can you describe in the pr description use case for that?

Added it.

**Comment by NikVolf** [comment]
So the end-user usecase will be that you show that gas cannot be calculated or what?

Add this as well maybe

**Comment by shamilsan** [comment]
> So the end-user usecase will be that you show that gas cannot be calculated or what?

Yes at the moment. But there are potential use cases in the future.

> Add this as well maybe

Done.

**Comment by breathx** [comment]
@NikVolf one more case for future is to forbid usage of sending syscalls in signals handling #647

**Review by breathx — CHANGES_REQUESTED** [review]
Research that moment with overriding host functions, please, and merge `master`.

**Comment by breathx** [comment]
Otherwise, great!

File: core-processor/src/configs.rs [hunk 1]
```diff
@@ -73,6 +74,8 @@ pub struct ExecutionSettings {
     pub existential_deposit: u128,
     /// Weights of host functions.
     pub host_fn_weights: HostFnWeights,
+    /// Functions forbidden to be called.
+    pub forbidden_funcs: Vec<&'static str>,
```
[the remark this answers is not in the document]
Comment (shamilsan) [reply core-processor/src/configs.rs]: Fixed

[the remark this answers is not in the document]
Comment (shamilsan) [reply core-processor/src/configs.rs]: Fixed

File: core-backend/sandbox/src/env.rs [hunk 2]
```diff
@@ -159,6 +147,26 @@ impl<E: Ext + IntoExtInfo + 'static> Environment<E> for SandboxEnvironment<E> {
         env_builder.add_host_func("env", "gr_wake", funcs::wake);
         env_builder.add_host_func("env", "gas", funcs::gas);
 
+        // Override forbidden functions by the dedicated handler.
+        ext.forbidden_funcs()
+            .iter()
+            .for_each(|f| env_builder.add_host_func("env", *f, funcs::forbidden));
```
[the remark this answers is not in the document]
Comment (shamilsan) [reply core-backend/sandbox/src/env.rs]: Fixed

File: core/src/env.rs [hunk 3]
```diff
@@ -261,8 +264,8 @@ mod tests {
     }
 
     /// Struct with internal value to interact with ExtCarrier
-    #[derive(Debug, PartialEq, Eq, Clone, Copy)]
-    struct ExtImplementedStruct(u8);
+    #[derive(Debug, PartialEq, Eq, Clone, Default)] //, Copy)]
+    struct ExtImplementedStruct(u8, BTreeSet<&'static str>);
```
[the remark this answers is not in the document]
Comment (breathx) [reply core/src/env.rs:268]: Seems useless

File: core/src/env.rs [hunk 4]
```diff
@@ -261,8 +264,8 @@ mod tests {
     }
 
     /// Struct with internal value to interact with ExtCarrier
-    #[derive(Debug, PartialEq, Eq, Clone, Copy)]
-    struct ExtImplementedStruct(u8);
+    #[derive(Debug, PartialEq, Eq, Clone, Default)] //, Copy)]
```
Comment (breathx) [inline core/src/env.rs:267]:
```suggestion
    #[derive(Debug, PartialEq, Eq, Clone, Default)]
```

dead code

File: core-backend/sandbox/src/env.rs [hunk 5]
```diff
@@ -71,6 +71,33 @@ pub(crate) struct Runtime<E: Ext> {
...
+}
+
+impl<'a, E: Ext + IntoExtInfo + 'static> EnvBuilder<'a, E> {
+    fn add_func(&mut self, name: &str, f: HostFuncType<Runtime<E>>)
+    where
+        E::Error: AsTerminationReason + IntoExtError,
+    {
+        if self.forbidden_funcs.contains(name) {
+            self.env_def_builder
+                .add_host_func("env", name, crate::funcs::FuncsHandler::forbidden);
```
Comment (breathx) [inline core-backend/sandbox/src/env.rs:88]:
```suggestion
                .add_host_func("env", name, FuncsHandler::forbidden);
```

Move import on top of the file

**Review by breathx — APPROVED** [review]

File: core-backend/sandbox/src/env.rs (hunk 5 above)
Comment (shamilsan) [reply core-backend/sandbox/src/env.rs]: Yes, my fault, fixed.

File: core/src/env.rs (hunk 4 above)
Comment (shamilsan) [reply core/src/env.rs]: Blind eye, repaired.

**Pull request merged**